corehq/apps/reports/views.py

```diff
@@ -1949,7 +1949,7 @@ def _get_case_link(case_id, name):
     ])
     msg = _("""Form cannot be archived as it creates cases that are updated by other forms.
         All other forms for these cases must be archived first:""")
-    notify_msg = """{} {}""".format(msg, case_links)
+    notify_msg = format_html("""{} {}""".format(msg, case_links))
```

Comment:
so `format_html` and `mark_safe` are kind of complementary -- `format_html` is used to combine 'safe' chunks, or make those chunks safe, whereas `mark_safe` labels chunks as safe. As written, `format_html` is going to receive what it perceives to be unsafe -- the message, followed by a bunch of html links. It will then escape all the HTML characters in them, because it is unsafe. Ideally, what you want is for `case_links` to be safe, which would then allow you to do:
```
format_html("{} {}", msg, safe_case_links)
```
The above will escape `msg`, which doesn't matter, because msg has no HTML. It will leave `safe_case_links` alone, because presumably it's been escaped/marked_safe earlier in our code.

The trick is making `case_links` safe earlier in the code. I would recommend using `format_html` for the initial link creation, and then `format_html_join` for combining those links.

Reply:
To add to what @mjriley said (which I agree with), I'd suggest you review every line in `_get_cases_with_forms_message` and use the right kind of escaping for each situation. For example,

```
return _("%(case_name)s (this case)") % {'case_name': name}
```
should use `format_html` to properly handle the values in `name`, and

```
'<a href="{}#!history">{}</a>'.format(reverse('case_data', args=[domain, case_id]), name)
```
should use `format_html` to properly escape the URL. (It doesn't matter that maybe we don't have to because we can show some other way that they'll never include `<`, `>`, etc. It's conceptually the right thing to do, and if indeed none of those characters are included it has no effect one way or the other. What it does do is let someone reading it know without any doubt that the end result is correctly escaped.)

Reply:
Thank you @dannyroberts @mjriley for your suggestions.

I have updated the code accordingly. 

I looked out for ways in which the `_get_cases_with_forms_message` can use `format_html_join`. I felt that it would be better not to use it. Since we are creating URLs with `reverse` that is taking variables `domain` and `case_id`. We also have different formats for the same case and the rest of the cases. 

So I went with Danny's suggestion and ensured that `format_html` handles all the variables that are passed in the function.

Let me know if you can think of a way in which the current implementation can be improved.